### DISCUSSION TITLE
OMCI_SW_VER* not sticking after reboot

### DISCUSSION THREAD

**Discussion by crrodriguez** [Anime4000/RTL960x#29] [Q&A]
I got an ODI DFP-34X-2C2 just today in the mail from fibershow (to see if it can replace carlitoxxpro ) . It seems you are indeed right that this stick is better designed and produces way less heat.. stick is now running firmware V1.0-220414 , but came with V1.0-220304, the same result occurs with both versions.

Even though I have supplied  the required parameters,

OMCI_OLT_MODE 3
PON_VENDOR_ID
GPON_ONU_MODEL
HW_HWVER
OMCI_SW_VER1
OMCI_SW_VER2

OMCI_SW_VER* is lost on reboot.. Now stuck with that.. it resets itself to default.. anyone seeing this ? in case it matters fiber is not plugged.

**Comment by vuducdong** [comment]
While you set OMCI_OLT_MODE is 3, it's mean need set CUSTOM_OMCI_SW_VER1 and CUSTOM_OMCI_SW_VER2 and maybe should try OMCC_VER

flash get OMCC_VER
OMCC_VER=128  --> 0x80

flash set OMCC_VER 134
OMCC_VER=134  --> 0x86

OMCC list (Hex):  0x80  0x81 0x82  0x83  0x84  0x85  0x86  0x96  0xA0  0xA1 0xB0  0xB1
            
OMCC list (Deci):  128  129   130   131   132   133   134   150   160  161   176  177

**Reply by crrodriguez** [reply]
there is no  CUSTOM_OMCI_SW_VER1 on this firmware, I already set the OMCC_VER equal to the isp crapbox

**Comment by Anime4000** [comment]
if `OMCI_SW_VER1` and `OMCI_SW_VER2` didn't work, try this last method `nv setenv sw_version0` and `nv setenv sw_version1` then reboot

```
# nv getenv sw_version0
sw_version0=V1.0-220304
# nv getenv sw_version1
sw_version1=V1.0-220414
```

modifying `nv` value is not recommended, can be override by firmware update, but can be use if `flash` didn't work

**Reply by crrodriguez** [reply]
ok, thanks, did that, sw_version* shows myu values after reboot, web interface doesn't . not sure what will prevail.. I'll let you to know.

**Reply by crrodriguez** [reply]
if it matters the value it is stuck on is V1.7.1 . not sure what component version is that.

**Reply by crrodriguez** [reply]
btw.. do you know how to contact ODI tech support ? couldnt find an english website or email..

**Comment by tdmadam** [comment]
During startup, the script chk_swver_2.sh compares sw_version0 with the version that is found in /etc/version
```
V1.0-220304 -- Fri Mar 4 11:22:18 CST 2022
```


if they don't match, it overwrites the nv variable with V1.0-220304

Probably fastest way would be to unpack rootfs, change V1.0-220304 -- to Sw0Example --
pack it, fix the md5 and flash modified software

or maybe disable chk_swver_2.sh in /etc/init.d/rc3

**Reply by crrodriguez** [reply]
but OMCI_SW_VER* gets overwritten to v1.7.1..

**Reply by crrodriguez** [reply]
this v1.7.1 string is hardcoded into libmib.so as default ...

**Comment by tdmadam** [comment]
What is your output of:
`nv getenv | grep sw`

Mine shows V1.0-220304 as software version after reboot wchich matches the /etc/version

**Reply by crrodriguez** [reply]
Yes, this doesd not seem to be the value used for omci..

**Comment by Anime4000** [comment]
What SW Version do you want to change? I try modify `rootfs` and special firmware for you

**Comment by tdmadam** [comment]
As a workaround, change OMCI_SW_VER1, OMCI_SW_VER2 to the desired values while the fiber optic cable is unplugged.
Plug the fiber optic cable back in and allow it to register with correct software version.

This behavior occurs when OMCI_OLT_MODE = 3

**Reply by crrodriguez** [reply]
I have the fiber unplugged, If I understood you comment..  OMCI_SW_VER* will only stick after the fiber was plugged ? if it will not stick after then your woraround does not help me. as I have to move the stick out of media converter and into the router where i cannot access the device from the web or telnet..

**Reply by tdmadam** [reply]
It will not stick if you move to another device, the OMCI_SW_VER1 and OMCI_SW_VER2 will get stored to /var/config/lastgood.xml but on reboot the default values V1.7.1 will be used if OMCI_OLT_MODE=3. Maybe something else needs to be changed to make it read the config, but I am not aware of it.

If ODI cannot fix it the other option is switch back to OMCI_OLT_MODE 0 and modify the firmware ourselves(etc/version).

**Reply by Anime4000** [reply]
as @tdmadam suggest, modify the firmware then edit `/etc/version` (

I have make Linux (Ubuntu) script to modify the firmware:
https://github.com/Anime4000/RTL960x/tree/main/Tools/emulator

* download `qemu-test.sh`
* edit `qemu-test.sh`
    * Go to [line number 95](https://github.com/Anime4000/RTL960x/blob/main/Tools/emulator/qemu-test.sh#L95)
    * Edit `STICKVER=`, example `STICKVER=168D.A` (Huawei)
    * Go to [line number 99](https://github.com/Anime4000/RTL960x/blob/main/Tools/emulator/qemu-test.sh#L99)
    * Edit `STICKVER=`, example `STICKVER=168D.A` (Huawei)
* open terminal, execute `sudo bash qemu-test.sh M114_sfp_ODI_220304.tar` (make sure script and firmware in same folder)
* once you see a `busybox`, simply type `exit`
* wait... script will repack and change `fwu_ver` and `/etc/version` value as you edit earlier
* new modified firmware will be created `rtl960x_modified.tar`
* update using `rtl960x_modified.tar`

You may need update twice for both mtd image partition

NOTE:
Editing `rootfs` will not work on WSL (Windows Subsystem Linux) because NTFS cannot emulate device file, must done in Linux

**Reply by tdmadam** [reply]
@Anime4000 
Works like a charm! Just need a little adjustement, that's what I get `168D.A-220507`

Edit:
Modified line 96. It shows now `168D.A`

![168DA](https://user-images.githubusercontent.com/52431348/167272173-1a198f12-67b3-4857-b177-54bfbe4c0c93.png)

**Reply by Anime4000** [reply]
nice~

to remove date build,

Line 96
```
echo "168D.A" > "$CHDIR/etc/version"
```
Line 97
```
echo "168D.A" > fwu_ver
```
Line 100
```
echo "168D.A" > "$CHDIR/etc/version"
```
Line 101
```
echo "168D.A" > fwu_ver
```

I need update the script, probably add extra parameter like
```
sudo bash qemu-test.sh M114_sfp_ODI_220304.tar 168D.A
```

**Reply by tdmadam** [reply]
Love this tool! Thanks!

**Reply by Anime4000** [reply]
@tdmadam Thanks! I made this because trying to port V2801F to TWCGPON657, it's proven useful, I use this to edit WebGUI

**Reply by Anime4000** [reply]
@tdmadam I have make quick update, new syntax is `sudo bash qemu-test.sh firmware.tar V5R019C00S125`

**Reply by crrodriguez** [reply]
I tried your script.. somehow.. boa now aborts and the web interface stops functioning

boa
size of linux : 1359872
size of linux : 3932160
init_global_parm888::size of g_max_upload_size : 3932160
/var/config/topbar.gif  is none exiting
counter = 29687
Link UP
/var/config/topbar.gif  is none exiting
Aborted

**Reply by crrodriguez** [reply]
telnet works however.

**Reply by tdmadam** [reply]
Yes, I have a similar problem with the second version of the script. The web interface crashes.
When I telnet, the /etc/version file contains

`V5R019C00S688`

with the previous, working version in /etc/version that I had

`V5R019C00S688 -- Sat May 07 23:10:50 UTC 2022`

Edit:
For my issue with "Using custom version string..." I replaced line 119

`echo "$2" > "$CHDIR/etc/version"`


with `echo "$2 -- $(date -u +'%a %b %d %H:%I:%M %Z %Y')" > "$CHDIR/etc/version"`

The web interface is up again.

**Reply by Anime4000** [reply]
I forgot to put `--`, because rc script looking a string before `--`
I have update the script, now easily to hardcode sw_version

**Reply by aoeII** [reply]
Hi Ainme, thank you for your amazing job, and I have followed your instruction to modify the software version of my ODI DFP-34X-2C2 stick to "3FE55630EFDB03" on the newest firmvare of "v1.0-220414" and flash the modified firmware to image-2. **Wonderful! It works!!!** However, when I reboot it , the OMCI software version on http://192.168.1.1/omci_info.asp would be reset to the hardcode of "V1.7.1". Could you please tell me how to update the OMCI software version permanently? 

When I flashed the the modified firmware to image-2 and conducted the first boot:
![0-1](https://user-images.githubusercontent.com/79325762/169010862-c9925588-9bda-415c-a993-8e39f6c148a7.png)
![0-2](https://user-images.githubusercontent.com/79325762/169010891-efcc5dd6-2835-448c-84bc-0916e6544aa7.png)

Please ignore the software version info of image-1, as I do not modify the v1.0-220304 firmware of image-1. I just use the command of "flash set OMCI_SW_VER1 3FE55630ADBC42" to update it, and it should be reset on the next boot.

After I rebooted the stick, it looked like this:
![1-1](https://user-images.githubusercontent.com/79325762/169012472-0935d90b-0e8c-47fb-b0b7-d1cd375f3ad0.png)
![1-2](https://user-images.githubusercontent.com/79325762/169012490-3815d99a-b977-4fef-a405-bfb8e9464ad7.png)
![1-3](https://user-images.githubusercontent.com/79325762/169012513-d35c9c61-8adc-442f-9d0f-2fd9e5907925.png)

Thank you for your patience.

> as @tdmadam suggest, modify the firmware then edit `/etc/version` (
> 
> I have make Linux (Ubuntu) script to modify the firmware: https://github.com/Anime4000/RTL960x/tree/main/Tools/emulator
> 
> * download `qemu-test.sh`
> * edit `qemu-test.sh`
>   
>   * Go to [line number 95](https://github.com/Anime4000/RTL960x/blob/main/Tools/emulator/qemu-test.sh#L95)
>   * Edit `STICKVER=`, example `STICKVER=168D.A` (Huawei)
>   * Go to [line number 99](https://github.com/Anime4000/RTL960x/blob/main/Tools/emulator/qemu-test.sh#L99)
>   * Edit `STICKVER=`, example `STICKVER=168D.A` (Huawei)
> * open terminal, execute `sudo bash qemu-test.sh M114_sfp_ODI_220304.tar` (make sure script and firmware in same folder)
> * once you see a `busybox`, simply type `exit`
> * wait... script will repack and change `fwu_ver` and `/etc/version` value as you edit earlier
> * new modified firmware will be created `rtl960x_modified.tar`
> * update using `rtl960x_modified.tar`
> 
> You may need update twice for both mtd image partition
> 
> NOTE: Editing `rootfs` will not work on WSL (Windows Subsystem Linux) because NTFS cannot emulate device file, must done in Linux

**Reply by rollingafull** [reply]
> * o to [line number 95](https://github.com/Anime4000/RTL960x/blob/main/Tools/emulator/qemu-test.sh#L95)

Hi,

The new lines to modify 'STICKVER=' are 105 and 109?

@Anime4000 what do you think about insert this explanation and/or @crrodriguez @stich86 explanation into and .md file like 'howtoo modify sw_version' into emulator section.

Good job guys.

Regards.

**Reply by Anime4000** [reply]
yea, but I plan to make modified firmware that allow change `swver` by using file `echo "V1abcef" > /etc/config/swver` which is safer than touching `nv` and `flash` value

**Comment by crrodriguez** [comment]
I ended up modifing /etc/insdrv.sh  so even before the drivers are loaded I have proper omci software version set. there i almost certainly a nicer way.. there must be some xml confuigursation that is always considered at boot.. I ran out of time on my investigation.

**Reply by aoeII** [reply]
> I ended up modifing /etc/insdrv.sh so even before the drivers are loaded I have proper omci software version set. there i almost certainly a nicer way.. there must be some xml confuigursation that is always considered at boot.. I ran out of time on my investigation.

Hi crrodriguez, thank you for your quick reply. Maybe that is not a perfect solution, but more than nothing, isn't it? Would you please share your modified /etc/insdrv.sh for my reference? Many thanks!  BTW, there seems no vi or vim in the shell environment of the stick, and I wonder how I can modify the batch command files? When I tried the command 'cat>filename.txt<<EOF' to creat a new file, the bash responsed: 'can't create filename.txt: Read-only file system'.

Thank you for your kindness.

**Reply by crrodriguez** [reply]
just open your favouirte editor in the squash-rootfs directory while it is unpacked..  just add at the top
flash set OMCI_SW_VER1 yourfirstswversion
flash set OMCI_SW_VER2 yoursecondswversion

NOW it also appears that modifing config_default_hs.xml adding valid xml atrtributes it should work..but I have not tested.

**Reply by aoeII** [reply]
> just open your favouirte editor in the squash-rootfs directory while it is unpacked.. just add at the top flash set OMCI_SW_VER1 yourfirstswversion flash set OMCI_SW_VER2 yoursecondswversion
> 
> NOW it also appears that modifing config_default_hs.xml adding valid xml atrtributes it should work..but I have not tested.

Oh, it sounds great! I will try it when I understand how to pack and unpack squash-rootfs system.
Thanks for your kind advice.

**Reply by aoeII** [reply]
> just open your favouirte editor in the squash-rootfs directory while it is unpacked.. just add at the top flash set OMCI_SW_VER1 yourfirstswversion flash set OMCI_SW_VER2 yoursecondswversion
> 
> NOW it also appears that modifing config_default_hs.xml adding valid xml atrtributes it should work..but I have not tested.

Thanks for your sugestion.

I made it! When I conducted the Ainme's script to modify `/etc/version`, there was a busybox promopt waiting for type `exit`. Then I shifted to another shell windows to excute `sudo vim the-path-to-working-dir/M114_sfp_ODi_vlan_220414/squashfs-root/etc/insdrv.sh`, and added the two lines at the top below `#!/bin/sh`, and type `:wq` to save it to harddisk.
```
flash set OMCI_SW_VER1 myswversion
flash set OMCI_SW_VER2 myswversion
```
Then I  shifted to the Ainme's script shell window and typed `exit` to continue. So cool! There was no error and acheived the modified firmware.
Flashed it to my ODI stick and rebooted twice to test it. It works! The `V1.7.1` disappeared without trace. I got the correct software version.

![2-1](https://user-images.githubusercontent.com/79325762/169191649-0d7483ae-ea21-4867-a421-e5c3d4ededcf.png)
![2-2](https://user-images.githubusercontent.com/79325762/169191677-d706923d-3d2f-4413-9d99-b358043d7aaa.png)
![2-3](https://user-images.githubusercontent.com/79325762/169191684-aafddd7d-4682-49ca-a6a7-1676af34eac5.png)

Thank you again. Thank Ainme. Thank you all.

**Reply by dolphin22** [reply]
> I ended up modifing /etc/insdrv.sh so even before the drivers are loaded I have proper omci software version set. there i almost certainly a nicer way.. there must be some xml confuigursation that is always considered at boot.. I ran out of time on my investigation.

How did you modify insdrv.sh? mine is read-only mode. Telnet and UART doesn't require login so I supposed root account.

**Comment by stich86** [comment]
following @crrodriguez suggestion, i've modified the `/etc/insdrv.sh` adding these lines after  `pon_mode=flash get PON_MODE`:

```
olt_mode=`flash get OMCI_OLT_MODE`

if [ "$olt_mode" == "OMCI_OLT_MODE=3" ]; then
        SWVER0=$(/bin/nv getenv sw_custom_version0 | awk -F'=' '{print $2}')
        SWVER1=$(/bin/nv getenv sw_custom_version1 | awk -F'=' '{print $2}')
        flash set OMCI_SW_VER1 $SWVER0
        flash set OMCI_SW_VER2 $SWVER1
fi
```

Then i've repacked the new image. 
Before update to the  modded firmware is needed to add two new variables on the uBoot env running on telnet session these two commands that should match your desired firmware version:

```
nv setenv sw_custom_version0 YOURFIRSTSWVER
nv setenv sw_custom_version1 YOURSECONDSWVER
```

check with this command if variables were written into the bootloader env:

`nv getenv | grep sw_custom`

**Reply by aoeII** [reply]
Wow! It is a pefect resolution. Thank you.

**Comment by vuducdong** [comment]
hi @Anime4000 directly updating via web the default version from manufacturer 220304 to other versions (HW_HWVER and ELAN_MAC_ADDR by default, nothing changed), will I have any problem with mac_key?

**Reply by Anime4000** [reply]
hi @vuducdong 

ODI `MAC_KEY` only ties with `ELAN_MAC_ADDR`, you can change `HW_HWVER` freely

if you lost `MAC_KEY`, you only can use default mac address firmware build-in, problem when using multiple ODI stick in same SPF Host can cause issue

**Comment by stich86** [comment]
@Anime4000 can you add also fix on BW issue for LAN_SDS_MODE > 4?
here is my current script (added into rc34):

sleep 120
fix_bw=`cat /proc/lan_sds/lan_sds_cfg | cut -d "(" -f 1 | cut -d "=" -f 2 | sed 's/ //g'`
if [ $fix_bw == "4" ] || [ $fix_bw == 5 ] || [ $fix_bw == 6 ]; then
        echo 'Fix upload issue for HiSGMII...'
        /bin/diag bandwidth set egress port all rate 4194296
        /bin/diag bandwidth set ingress port all rate 4194296
        echo 'Done!'
fi

**Reply by Anime4000** [reply]
nice script,

but I don't have extra ODI Stick to try with, afraid my current stick out of order

**Reply by stich86** [reply]
> nice script,
> 
> 
> 
> but I don't have extra ODI Stick to try with, afraid my current stick out of order

Ive it on my stick ;)

**Comment by aoeII** [comment]
Done !

[M114_sfp_ODi_vlan_220414_rtl960x_modified_20220602.zip](https://github.com/Anime4000/RTL960x/files/8820036/M114_sfp_ODi_vlan_220414_rtl960x_modified_20220602.zip)
MD5 for the tar (not the zip) is: 810C5712A271AE86B9C4ED367D4683FA
It  includes [stich86](https://github.com/stich86) ‘s scripts for `/etc/insdrv.sh` and `/etc/init.d/rc34`. 

Thank you all.

![omci_mod_0602](https://user-images.githubusercontent.com/79325762/171541448-fcc4bf98-3b06-461e-a1ad-bf45aad79fb9.png)

~~B.T.W., I found an interesting web page on ODI DFP-34X-2C2 "http://192.168.1.1/vermod.asp",~~
~~It may be more choices to have fun.~~

 I am very sorry for that the OMCI Export function in `vermod.asp` would make the stick brick. So forget it please.

**Reply by Anime4000** [reply]
nice!

**Reply by vuducdong** [reply]
@Anime4000 and @aoeII  Good morning,

Have you ever had a situation with OLT Alcatel/Nokia, if change SFP's GPON SN to same as original ONT's GPON SN (from ISP), then can't access to sfp's web or telnet/ssh interface, unless you change it to be different from the original ONT, then access it normally.

**Reply by stich86** [reply]
nice finding, but looks like that vermod.asp is not present on 2205XX firmware.

I'm attaching same mods for the 2205XX firmware if anyone need it:

[M114_sfp_ODI_hybrid_220527_stich86_220530.tar.zip](https://github.com/Anime4000/RTL960x/files/8821809/M114_sfp_ODI_hybrid_220527_stich86_220530.tar.zip)

MD5 for the tar (not the zip) is: 8414bab811ce51ac91000828504c61f3

**Reply by aoeII** [reply]
> 
Hi stich, thanks for you amazing scripts.

As shown above, the link for`vermod.asp` is valid on 220414 firmware.  I have not flashed 2205XX firmware to my stick as Anime said it is not a bridge sfu firmware.

**Reply by aoeII** [reply]
> @Anime4000 and @aoeII Good morning,
> 
> Have you ever had a situation with OLT Alcalte/Nokia, if change SFP's GPON SN to same as original ONT's GPON SN (from ISP), then can't access to sfp's web or telnet/ssh interface, unless you change it to be different from the original ONT, then access it normally.

Sorry about it and I have no idea on it. Maybe the OLT shut the ONU's web access and telnet port by omci command for safety reason.

**Reply by crrodriguez** [reply]
I have the same problem, cannot telnet or web if such connection is initiated after the fiber is plugged and up, OLT ALCL

**Comment by ToTheCLI** [comment]
I tried both custom firmware versions and none are saving the value I entered aka does not survive reboot
Any other way?

**Reply by eugenidev** [reply]
Hi, get [M114_sfp_ODI_hybrid_220527_stich86_220530.tar.zip](https://github.com/Anime4000/RTL960x/files/8821809/M114_sfp_ODI_hybrid_220527_stich86_220530.tar.zip) and execute:
```
nv setenv sw_custom_version0 YOURFIRSTSWVER
nv setenv sw_custom_version1 YOURSECONDSWVER
```
![Screenshot_5](https://user-images.githubusercontent.com/22867605/176784672-6c865403-9494-4562-bc42-4a7217c9ef9c.png)
But It`s not resolve my global problem.

**Reply by stich86** [reply]
The changes works only when OLT MODE is 3, otherwise are no evaluated
